mongodb-collections-fragmented.md (#84)
percona-platform/portal-doc: docs/advisors/checks/mongodb-collections-fragmented.md

> # MongoDB Fragmented Collections

Reviewer: ```suggestion
# MongoDB fragmented collections
```

## Description

> This check returns a warning if the Storage size is greater than the Data size of a collection. That condition indicates that the collection is fragmented and needs a Compaction or Initial sync to reclaim disk space.

Reviewer: ```suggestion
This check warns if the storage size is greater than the data size of a collection. That condition indicates that the collection is fragmented and needs a compaction or initial sync to reclaim disk space.
```

## Resolution

When you are getting high disk usage and for a wiredTiger storage engine you observe that some of the collections have storage size greater than the data size, then it means that those collections are fragmented. 

This often happens if your application deletes many documents in your collections. The WiredTiger storage engine maintains lists of empty records in data files as it deletes documents. This space can be reused by WiredTiger, but will not be returned to the operating system. 

> In order to reclaim the disk space, there are 2 ways -

Reviewer: ```suggestion
In order to reclaim the disk space, there are 2 ways:
```

- Run compact on the collections
- Resync the node

**Run compact on the collections:**

> The Compact command rewrites and defragments all data and indexes in a collection. On WiredTiger databases, this command will release unneeded disk space to the operating system.

Reviewer: ```suggestion
The `compact` command rewrites and defragments all data and indexes in a collection. On WiredTiger databases, this command will release unneeded disk space to the operating system.
```

Below is the syntax -
> db.runCommand({compact: _collection name_})

**NOTE:**

>   Starting in MongoDB 6.0.2 (and 5.0.12, and 4.4.17):

Reviewer: ```suggestion
  Starting in MongoDB 6.0.2 (and 5.0.12, and 4.4.17):
  
```

- A secondary node can replicate while compact is running.
- Reads are permitted.
- All other operations are permitted, except the below ones -
  - db.collection.drop()
  - db.collection.createIndex()
  - db.collection.createIndexes()
  - db.collection.dropIndex()
  - db.collection.dropIndexes()
  - collMod

> **Important:**
> 
>  Always have an up-to-date backup before performing server maintenance such as the compact operation.
> 

> > For more details on Compact command, kindly refer to [MongoDB Documentation](https://www.mongodb.com/docs/manual/reference/command/compact/).

Reviewer: ```suggestion
> For more details on the `compact` command, refer to [MongoDB Documentation](https://www.mongodb.com/docs/manual/reference/command/compact/).
```

  
**Resync the node:**
  

> Instead of running compact command on a collection, you can resync the node in a rolling fashion so that there will be no downtime or impact on the application.

Reviewer: ```suggestion
Instead of running the `compact` command on a collection, you can resync the node in a rolling fashion so that there will be no downtime nor impact on the application.
```

> This approach is the safest way to reclaim the disk space but it can be time consuming if your Data size is huge.

Reviewer: ```suggestion
This approach is the safest way to reclaim the disk space but it can be time consuming if your data size is huge.
```

# MongoDB fragmented collections

## Description
This check warns if the storage size is greater than the data size of a collection. This means that the collection is fragmented and needs running a compaction or an initial sync to reclaim disk space.

## Resolution

> High disk usage and wiredTiger storage engines with collections that have storage sizes greater than data size indicate that these collections are fragmented. 

Author: Here it is trying to say that,
1. When you are getting High Disk usage and
2. You observe that, for a WiredTiger storage engine, some collections have storage size greater than the data size.

Then, it means that those collections are fragmented.

> High disk usage and wiredTiger storage engines with collections that have storage sizes greater than data size indicate that these collections are fragmented. 

Reviewer 2: ```suggestion
Fragmented collections are indicated by:
- wiredTiger storage engines with collections that have storage sizes greater than data size AND
- high disk usage```

This often happens if your application deletes many documents in your collections. The WiredTiger storage engine maintains lists of empty records in data files as it deletes documents. This space can be reused by WiredTiger, but will not be returned to the operating system. 

> In order to reclaim the disk space, there are 2 ways:

Reviewer 2: ```suggestion
To reclaim disk space, either:
```

- Run compact on the collections
- Resync the node

**Run compact on the collections:**

The `compact` command rewrites and defragments all data and indexes in a collection. On WiredTiger databases, this command releases unneeded disk space to the operating system.

Below is the syntax -
> db.runCommand({compact: _collection name_})

**NOTE:**

  Starting in MongoDB 6.0.2 (and 5.0.12, and 4.4.17):
  
- A secondary node can replicate while compact is running.
- Reads are permitted.
- All other operations are permitted, except the below ones -
  - db.collection.drop()
  - db.collection.createIndex()
  - db.collection.createIndexes()
  - db.collection.dropIndex()
  - db.collection.dropIndexes()
  - collMod

> **Important:**
> 
>  Always have an up-to-date backup before performing server maintenance such as the compact operation.
> 

> > For more details on the `compact` command, refer to [MongoDB Documentation](https://www.mongodb.com/docs/manual/reference/command/compact/).

Reviewer 2: ```suggestion
> For more details on the `compact` command, see the [MongoDB Documentation](https://www.mongodb.com/docs/manual/reference/command/compact/).
```

  
**Resync the node:**
  
Instead of running the `compact` command on a collection, you can resync the node in a rolling fashion so that there will be no downtime nor impact on the application.

This approach is the safest way to reclaim the disk space but it can be time consuming if your data size is huge.

> For more information on how to reclaim disk space, see the [Percona Blog](https://www.percona.com/blog/how-to-reclaim-disk-space-in-percona-server-for-mongodb/).

Reviewer 2: ```suggestion
For more information on reclaiming disk space, see the [Percona Blog](https://www.percona.com/blog/how-to-reclaim-disk-space-in-percona-server-for-mongodb/).
```
